src/toil/resource.py

```diff
@@ -181,27 +196,27 @@ def localPath(self):
         raise NotImplementedError
 
     @property
-    def localDirPath(self):
+    def localDirPath(self) -> str:
         """
         The path to the directory containing the resource on the worker.
         """
         rootDirPath = os.environ[self.rootDirPathEnvName]
         return os.path.join(rootDirPath, self.contentHash)
 
-    def pickle(self):
+    def pickle(self) -> str:
         return self.__class__.__module__ + "." + self.__class__.__name__ + ':' + json.dumps(self)
 
     @classmethod
-    def unpickle(cls, s) -> "Resource":
+    def unpickle(cls, s: str) -> "Resource":
         className, _json = s.split(':', 1)
-        return locate(className)(*json.loads(_json))
+        return locate(className)(*json.loads(_json)) # type: ignore
 
     @classmethod
-    def _pathHash(cls, path):
+    def _pathHash(cls, path: str) -> str:
         return hashlib.md5(path.encode('utf-8')).hexdigest()
 
     @classmethod
-    def _load(cls, path):
+    def _load(cls, path: str) -> Union[IO[bytes], IO[str]]:
```

Comment:
This doesn't look like a useful type for this to be. This is the same `_load` I was complaining about above where you only use the result if it is a `bytes`, right?

I don't think discarding `str` versions is the right approach; we should make this always `IO[bytes]` and then anything that isn't currently returning an `IO[bytes]` probably should be changed to do so.

Reply:
Also you can drop all the `:rtype:` comments now that we have real type hints.